### ISSUE TITLE
[Core] ray memory CLI hang if any actor hangs.

### ISSUE THREAD

**Issue by MissiontoMars** [ray-project/ray#41431]
### What happened + What you expected to happen

There are two issues in the reproduction script below:

1. Holding the GIL lock in MasterActor causes ray.get in WorkerC to hang.
2. Attempting to use the "ray memory" command to check the status of ObjectRef hangs.

Both of these issues seem to be caused by the IO thread of MasterActor unable to acquire the GIL lock.

### Versions / Dependencies

ray2.8.0
But I believe the latest master branch should also have this issue.

### Reproduction script

```
import ray
import time

ray.init()

@ray.remote
class WorkerA:
    def __init__(self):
        pass

    def do_somthing(self):
        import time
        time.sleep(5)
        return 100

@ray.remote
class WorkerB:
    def __init__(self):
        pass

    def do_somthing(self):
        import time
        time.sleep(5)
        return 100

@ray.remote
class WorkerC:
    def __init__(self):
        pass

    def do_somthing(self, refs):
        print(ray.get(refs[0]))

@ray.remote
class MasterActor:
    def __init__(self):
        self.worker_a = WorkerA.remote()
        self.worker_b = WorkerB.remote()
        self.worker_c = WorkerC.remote()

    async def train(self):
        print('MasterActor train enter')
        ref_a = self.worker_a.do_somthing.remote()
        ref_b = self.worker_b.do_somthing.remote()
        self.worker_c.do_somthing.remote([ref_b])
        await ref_a
        print('MasterActor train leave')

    def hang(self):
        print('hang enter')
        # Holding gil for a long time.
        import re
        re.match(r'(a?){1000}a{1000}', 'a'*1000)
        print('hang leave')

trainer_actor = MasterActor.remote()
trainer_actor.train.remote()
time.sleep(2)
trainer_actor.hang.remote()

import time
time.sleep(100000)
```

ray memory：

(ray) ➜  logs ray memory
Traceback (most recent call last):
  File "/Users/wanxing/miniforge3/envs/ray/bin/ray", line 33, in <module>
    sys.exit(load_entry_point('bytedray', 'console_scripts', 'ray')())
  File "/Users/wanxing/work/code/ray/python/ray/scripts/scripts.py", line 2497, in main
    return cli()
  File "/Users/wanxing/miniforge3/envs/ray/lib/python3.9/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/Users/wanxing/miniforge3/envs/ray/lib/python3.9/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/Users/wanxing/miniforge3/envs/ray/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/wanxing/miniforge3/envs/ray/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/wanxing/miniforge3/envs/ray/lib/python3.9/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/Users/wanxing/work/code/ray/python/ray/scripts/scripts.py", line 1947, in memory
    mem_stats = memory_summary(
  File "/Users/wanxing/work/code/ray/python/ray/_private/internal_api.py", line 43, in memory_summary
    return memory_summary(
  File "/Users/wanxing/work/code/ray/python/ray/dashboard/memory_utils.py", line 399, in memory_summary
    node_stats(raylet["NodeManagerAddress"], raylet["NodeManagerPort"])
  File "/Users/wanxing/work/code/ray/python/ray/_private/internal_api.py", line 104, in node_stats
    node_stats = stub.GetNodeStats(
  File "/Users/wanxing/miniforge3/envs/ray/lib/python3.9/site-packages/grpc/_channel.py", line 946, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/Users/wanxing/miniforge3/envs/ray/lib/python3.9/site-packages/grpc/_channel.py", line 849, in _end_unary_response_blocking
    raise _InactiveRpcError(state)
grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.DEADLINE_EXCEEDED
	details = "Deadline Exceeded"
	debug_error_string = "UNKNOWN:Deadline Exceeded {created_time:"2023-11-28T11:30:50.222337+08:00", grpc_status:4}"

### Issue Severity

None

**Comment by rkooo567** [comment]
cc @c21 this may be related to issue you observe. We will take a look at it shortly

**Comment by c21** [comment]
Thank you @rkooo567!

**Comment by ruisearch42** [comment]
The detailed steps that cause `WorkerC.do_something()` to hang is as follows:

- When `WorkerC` executes `do_something()` with `execute_task()` (_raylet.pyx) , it deserializes its arguments `refs` with `ray._private.worker.global_worker.deserialize_objects()`
- Since the ObjectRef in `refs` is not resolved yet, `deserialize_objects()` makes an RPC call `GetObjectStatus` to the ObjectRef’s owner `MasterActor`, and execution blocks until `MasterActor` replies with the resolved object
- `MasterActor` runs `train()` and blocks at `await ref_a`, which calls `CoreWorker.GetAsync()` with `async_callback()` (in _raylet.pyx, requires GIL) as a callback
- Since `MasterActor` is an async actor, `hang()` starts to execute before `train()` finishes. The method `hang()` calls into `re.match()` which has a known issue that it holds the GIL and does not release it: https://bugs.python.org/issue23690 
- `WorkerA` finishes executing `do_something()`, and `MasterActor` puts the object (corresponding to `ref_a`) in memory store as part of `HandlePushTaskReply` 
- However, since `async_callback()` needs to acquire GIL to continue execution, which is held forever by `hang()` method in another thread, so it completely blocks the I/O thread
- This also blocks `MasterActor`’s reply to `WorkerC`’s `GetObjectStatus` RPC (which also runs on the I/O thread), even if `ref_b` is already resolved from the RPC call to `WorkerB`. As a result, `WorkerC` hangs

In summary, for an async actor, calling a method that holds GIL forever would block execution of all other Python threads, including I/O threads which communicate with other workers. 

@MissiontoMars Is it possible to not hold the GIL forever in `hang`?
